tests/test_clean.py

```diff
@@ -154,83 +269,168 @@ def test_bare_entities(text, expected):
+        {'tags': ['a', 'p'], 'strip': True},
         '<p><a href="http://example.com/"></a></p>'
-    )
+    ),
+])
+def test_stripping_tags(data, kwargs, expected):
+    assert clean(data, **kwargs) == expected
+
+
+@pytest.mark.parametrize('data, expected', [
+    (
+        '<scri<script>pt>alert(1)</scr</script>ipt>',
+        'pt&gt;alert(1)ipt&gt;'
+    ),
+    (
+        '<scri<scri<script>pt>pt>alert(1)</script>',
+        'pt&gt;pt&gt;alert(1)'
+    ),
+])
+def test_stripping_tags_is_safe(data, expected):
+    """Test stripping tags shouldn't result in malicious content"""
+    assert clean(data, strip=True) == expected
 
 
 def test_allowed_styles():
+    """Test allowed styles"""
     ATTRS = ['style']
     STYLE = ['color']
 
     assert (
-        bleach.clean('<b style="top:0"></b>', attributes=ATTRS) ==
+        clean('<b style="top:0"></b>', attributes=ATTRS) ==
         '<b style=""></b>'
     )
 
     text = '<b style="color: blue;"></b>'
-    assert bleach.clean(text, attributes=ATTRS, styles=STYLE) == text
+    assert clean(text, attributes=ATTRS, styles=STYLE) == text
 
     text = '<b style="top: 0; color: blue;"></b>'
     assert (
-        bleach.clean(text, attributes=ATTRS, styles=STYLE) ==
+        clean(text, attributes=ATTRS, styles=STYLE) ==
         '<b style="color: blue;"></b>'
     )
 
 
-def test_lowercase_html():
-    """We should output lowercase HTML."""
+def test_href_with_wrong_tag():
     assert (
-        bleach.clean('<EM CLASS="FOO">BAR</EM>', attributes=['class']) ==
-        '<em class="FOO">BAR</em>'
+        clean('<em href="fail">no link</em>') ==
+        '<em>no link</em>'
     )
 
 
+def test_disallowed_attr():
+    IMG = ['img', ]
+    IMG_ATTR = ['src']
+
+    assert (
+        clean('<a onclick="evil" href="test">test</a>') ==
+        '<a href="test">test</a>'
+    )
+    assert (
+        clean('<img onclick="evil" src="test" />', tags=IMG, attributes=IMG_ATTR) ==
+        '<img src="test">'
+    )
+    assert (
+        clean('<img href="invalid" src="test" />', tags=IMG, attributes=IMG_ATTR) ==
+        '<img src="test">'
+    )
+
+
+def test_unquoted_attr_values_are_quoted():
+    assert (
+        clean('<abbr title=mytitle>myabbr</abbr>') ==
+        '<abbr title="mytitle">myabbr</abbr>'
+    )
+
+
+def test_unquoted_event_handler_attr_value():
+    assert (
+        clean('<a href="http://xx.com" onclick=foo()>xx.com</a>') ==
+        '<a href="http://xx.com">xx.com</a>'
+    )
+
+
+def test_invalid_filter_attr():
+    IMG = ['img', ]
+    IMG_ATTR = {
+        'img': lambda tag, name, val: name == 'src' and val == "http://example.com/"
+    }
+
+    assert (
+        clean('<img onclick="evil" src="http://example.com/" />', tags=IMG, attributes=IMG_ATTR) ==
+        '<img src="http://example.com/">'
+    )
+    assert (
+        clean('<img onclick="evil" src="http://badhost.com/" />', tags=IMG, attributes=IMG_ATTR) ==
+        '<img>'
+    )
```

Comment:
This test existed elsewhere in the file. I moved it, but it is otherwise unchanged.